src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+            {
+                _outputLength = AnomalyOnlyOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndMargin))
+            {
+                env.CheckUserArg(sensitivity >= 0 && sensitivity <= 100, nameof(sensitivity), "Must be in [0,100].");
+                _outputLength = AnomalyAndMarginOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+            else if (detectMode.Equals(SrCnnDetectMode.AnomalyAndExpectedValue))
+            {
+                _outputLength = AnomalyAndExpectedValueOutputLength;
+                _modler = new SrCnnEntireModeler(threshold, sensitivity, detectMode, _outputLength);
+            }
+
+            SchemaBindings = new Bindings(input.Schema, inputColumnName, outputColumnName, new VectorDataViewType(NumberDataViewType.Double, _outputLength));
+        }
+
+        protected override ColumnBindingsBase SchemaBindings { get; }
+
+        protected override Delegate[] CreateGetters(DataViewRowCursor input, Batch currentBatch, bool[] active)
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, _inputColumnName) };
+        }
+
+        protected override Batch InitializeBatch(DataViewRowCursor input) => new Batch(_batchSize, _outputLength, _modler);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override ValueGetter<double> GetLookAheadGetter(DataViewRowCursor input)
+        {
+            return input.GetGetter<double>(input.Schema[_inputColumnName]);
+        }
+
+        protected override Func<int, bool> GetSchemaBindingDependencies(Func<int, bool> predicate)
+        {
+            return (SchemaBindings as Bindings).GetDependencies(predicate);
+        }
+
+        protected override void ProcessExample(Batch currentBatch, double currentInput)
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        public sealed class Batch
+        {
+            private List<double> _previousBatch;
+            private List<double> _batch;
+            private readonly int _batchSize;
+            private readonly int _outputLength;
+            private SrCnnEntireModeler _modler;
+            private double[][] _results;
```

Comment:
>_results [](start = 31, length = 8)

Why not make this readonly? That way you don't have to allocate a new array for every batch.

Reply:
Due to the scenario that bachSize=-1, and we do not know the exact row count until we read in data, results could not be allocated when initialize Batch.

---
In reply to: [426411747](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426411747) [](ancestors = 426411747)